sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs

```diff
@@ -0,0 +1,196 @@
+
+                    var totalClassCoupling = Enumerable.Empty<ITypeSymbol>()
+                        .Union(allFieldsCoupling)
+                        .Union(allPropertiesCoupling)
+                        .Union(allMethodsCoupling);
+                    var classCouplingCount = FilterAndCountCoupling(totalClassCoupling);
+
+                    if (classCouplingCount > Threshold)
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation(),
+                            classCouplingCount, Threshold));
+                    }
+                }, SyntaxKind.ClassDeclaration);
+        }
+
+        private static IEnumerable<ITypeSymbol> CollectCoupledClasses(FieldDeclarationSyntax field,
+            SemanticModel model)
+        {
+            return field.Declaration.Variables
+                .Select(variable => variable.Initializer != null
+                    ? model.GetTypeInfo(variable.Initializer.Value)
+                    : default(TypeInfo))
+                .SelectMany(x => new[] { x.Type, x.ConvertedType });
+        }
+
+        private static IEnumerable<ITypeSymbol> CollectCoupledClasses(BasePropertyDeclarationSyntax property,
+            SemanticModel model)
+        {
+            var propertyReturnType = new[] { model.GetSymbolInfo(property.Type).Symbol as ITypeSymbol };
+
+            var indexerProperty = property as IndexerDeclarationSyntax;
+            if (indexerProperty?.ExpressionBody != null)
+            {
+                // Arrowed property
+                return propertyReturnType
+                    .Union(indexerProperty.ExpressionBody
+                        .DescendantNodes()
+                        .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model)));
+            }
+
+            var basicProperty = property as PropertyDeclarationSyntax;
+            if (basicProperty?.ExpressionBody != null)
+            {
+                // Arrowed property
+                return propertyReturnType
+                    .Union(basicProperty.ExpressionBody
+                        .DescendantNodes()
+                        .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model)));
+            }
+
+            // Classic property OR event
+            return propertyReturnType
+                .Union(property.AccessorList.Accessors
+                    .SelectMany(accessor => accessor.DescendantNodes()
+                        .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model))));
+        }
+
+        private static IEnumerable<ITypeSymbol> CollectCoupledClasses(BaseMethodDeclarationSyntax method,
+            SemanticModel model)
+        {
+            return method.DescendantNodes()
+                .SelectMany(node => ExtractTypeSymbolsFromSyntaxNode(node, model));
+        }
+
+        private static IEnumerable<ITypeSymbol> ExtractTypeSymbolsFromSyntaxNode(SyntaxNode node, SemanticModel model)
```

Comment:
perhaps if you use `yield return` in this method the code will look better: less created arrays, no need to return null in some cases (e.g. when parameter.Type == null), etc.